Migrate to spdx-tools 0.8.0a2 (a pre-release)

jspeed-meyers opened the pull request:
Fix #113 
Fix #70 

This branch introduces v0.8.0a2 of spdx-tools. This migration required substantial changes to the codebase (for the better, honestly) and the test documents and test cases.

For those following along at home:

There is new version of `spdx-tools` (thanks to the great efforts of @meretp and others) and it creates breaking changes. This migration addresses those breaking changes. Importantly, some of those breaking changes also led to new test documents (the new `spdx-tools` parsing rules are stricter) and so that's why there are so many changes lines of code. For reviewers, the most important changes are in `sbom_checker.py`, IMO.

This is a larger change than I would prefer to make in a single PR. But I don't see any way around it short of having a broken build. This way we avoid a broken build and complete the migration.

@goneall, @meretp, @linynjosh -- Feel free to take a look over the next week. Assuming positive reviews and no objections, we can merge this into main in a week. :)

Again, big kudos to @meretp, without whom this migration would not have happened or would have happened at a glacially slow rate. THANK YOU!

Commit:
Update spdx-tools version to alpha release 0.8.0a1

spdx-tools implemented a significant refactor. This commit
bumps the version and is expected to introduce breaking changes
to ntia-conformance-checker.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Enable build test to use pre-release versions of packages

Because spdx-tools has been refactored and is available now only
in a pre-release versions, this change enables the build test
pipeline to use prerelease versions of packages, which is not
allowed by pip/pipenv with default settings.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix creation info and SpdxNoAssertion for migration spdx-tools 0.8.0a1

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

itsyongan reviewed (approved):
Thanks @jspeed-meyers!

jspeed-meyers commented:
Thanks, @linynjosh! I'll work to figure out the migration before merging :)

Commit:
Change import path to parse_anything function

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

itsyongan reviewed (approved):
Thanks @jspeed-meyers!

Commit:
Add missing date to a YAML SBOM

The spdx-tools refactoring made the spdx spec validation stricter
and this document no longer matches the spdx spec. Adding this
created timestamp makes the document compliant.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Change HTML test expected ouput

Since the new spdx-tools version has stricter requirements that the
SBOM comply with spdx spec, the SBOM used for this test needed to
be ammended and so now too does the expected test output need to
be ammended since there is a timestamp in the underlying SBOM.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Implement various migration changes

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Flip timestamp expected value in json output test

Because the underlying test SBOM now does have a timestamp,
because the refactored version of spdx-tools requires ingested
SBOMs to conform strictly to the spdx spec, this test now ought
to expect that an SBOM does have a timestamp.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix spdxnoassertion logic related to package supplier for migration

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Merge branch 'main' into migrate2spdx-tools-0.8.0a1

Commit:
Enable pre-release for testing this branch

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Update photon SBOM and drop version for one of the components

The migration to the 0.8.0 spdx-tools library version means
dealing with stricter adherence to the spdx spec. This commit
drops a version field that is empty since it was throwing an
error with the new library.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Implement pytest exception pattern for missing unique identifiers

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix issues with missing supplier test SBOMs

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix errors related to missing component version test SBOMs

Drop RDF SBOM since there seem to be so many specification-related
issues.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix pylint and black erros with sbom_checker.py

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

meretp commented on ntia_conformance_checker/sbom_checker.py:
```diff
@@ -37,28 +37,24 @@ def parse_file(self):
+                ActorType.PERSON,
+                ActorType.ORGANIZATION,
+                ActorType.TOOL,
+            ]:
```
The refactored version of `tools-python` already takes care of this check while parsing. If the document doesn't contain a valid creator or anything that can't be parsed to an `Actor` object a `SPDXParsingError` would be raised while parsing.

jspeed-meyers replied:
Ah, okay. Thank you. I will change the logic here too, saving myself/the codebase of managing this logic :)

meretp replied:
I had another look at the implementation in the `tools-python` (or `spdx-tools`, sorry for mixing the wording here) and have to admit that the tools would accept an empty list. Sorry for the misinformation. So we would need to check if `self.doc.creation_info.creators` contains at least one element here. For `self.doc.creation_info.created` my first comment is true and the parser would fail if no value is provided.

jspeed-meyers replied:
Ah, okay. There is very harmless "misinformation." So the logic should be:

```python
if len(self.doc.creation_info.creators) > 0:
```

jspeed-meyers replied:
Implemented this in [107c8b7](https://github.com/spdx/ntia-conformance-checker/pull/114/commits/107c8b758313e19b4f7225f782e94244b04a9d50)

meretp commented on ntia_conformance_checker/sbom_checker.py:
```diff
@@ -18,7 +18,7 @@ class SbomChecker:
+        self.sbom_name = self.doc.creation_info.name
         self.doc_version = self.check_doc_version()
         self.doc_author = self.check_doc_author()
         self.doc_timestamp = self.check_doc_timestamp()
```
These two checks are also taken care of by the parser. If the document doesn't contain a valid creator or created property a SPDXParsingError is raised from the `parse_anything.parse_file()` method as the file isn't a valid SPDX document. I am not sure if it makes sense to catch this Error to be able to generate a similar output as it is now.

jspeed-meyers replied:
Good point. I will remove these functions and just set these values equal to true :) Less code is (almost) always better. Thank you for saving me/this codebase from more code :)

jspeed-meyers replied:
Implemented this with [84e411c](https://github.com/spdx/ntia-conformance-checker/pull/114/commits/84e411cf43cdfaa6fb24b244b26f91f08730fa5e)

meretp reviewed (commented):
Thanks for opening this PR, the changes made look good to me and fit what I would have expected. 
I have two remarks on the current changes which are associated to some basic changes coming with the refacotred version. So to provide a bit more background on the refactored version of `tools-python`:
On the one hand we introduced a new separate layer for the validation of a parsed SPDX document. This validation takes care of the more "implicit rules for SPDX documents". So for example each `SPDXID` used in the document need to be unique. It might make sense to also use this validation here and expect a valid SPDX document before checking for the minimal ntia elements. 
On the other hand we introduced a more strict underlying model, which enforces correct types and cardinalities for each property. So for example an SPDX document must have at least one `creator` that is a person, a tool or an organization. These types of "validation" are handled by the parser and the tools will raise a `SPDXParsingError` if the parsed document doesn't fit the model. 

I think this is also the reason for most of the failing tests. I had a look at some of the test files and noticed that they aren't valid according to the spec. For example the [SPDXJsonExample.json](https://github.com/spdx/ntia-conformance-checker/blob/main/tests/data/no_elements_missing/SPDXJsonExample.json) contains a surrounding `Document` object which is not valid according to the spec.  I think a first step to fix the tests is to make all used documents valid SPDX documents. Apart from that I will take a closer look at the tests and try to provide more feedback on why they fail and how to fix them.

jspeed-meyers commented:
> It might make sense to also use this validation here and expect a valid SPDX document before checking for the minimal ntia elements.

Perfect. I like this.

> I think a first step to fix the tests is to make all used documents valid SPDX documents.

Good call. Will do.

> Apart from that I will take a closer look at the tests and try to provide more feedback on why they fail and how to fix them.

Thank you. You've really gone above and beyond in helping this project and others do the migration. THANK YOU!

meretp commented:
Your welcome! This also helps us a lot to verify that the refactored version is stable. I already found a small [bug](https://github.com/spdx/tools-python/issues/611) thanks to this PR, so a big thank you too for being willing to serve as an early adopter! :slightly_smiling_face:

Commit:
Simplify checking doc author by simply checking for non-zero number

Because the new spdx-tools version won't accept does the work of the
old check, that old code can be replaced by a much simpler version.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Simplify doc_author and doc_timestamp values to true

If the document is parsed successfully, the new spdx-tools logic
guarantees that there is an author and there is a document
timestamp. So this commit sets those values to always true.

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Fix test documents related to missing relationships

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

meretp commented on tests/data/missing_dependency_relationships/SPDXJsonExample.json:
```diff
@@ -1,4 +1,10 @@
+        "creators" : [ "Tool: LicenseFind-1.0", "Organization: ExampleCodeInspect ()", "Person: Jane Doe ()" ],
+        "licenseListVersion" : "3.17"
+      },
     "Document": {
```
This key needs to be deleted as it is no longer valid since 2.2 (see the [schema](https://github.com/spdx/spdx-spec/blob/development/v2.2.2/schemas/spdx-schema.json) and also the examples for [SPDX 2.2](https://github.com/spdx/spdx-spec/blob/development/v2.2.2/examples/SPDXJSONExample-v2.2.spdx.json) and [SPDX 2.3](https://github.com/spdx/spdx-spec/blob/development/v2.3.1/examples/SPDXJSONExample-v2.3.spdx.json). I would recommend to replace all test documents with the valid examples from the spec repo and then adjust them according to the test cases.

jspeed-meyers replied:
@meretp, good call. I'll do that. Thank you!

meretp commented:
FYI: We have a weekly call to discuss topics related to the `spdx-tools`. It's open to everyone, and meant as an informal way to sync on current topics and where development should focus next. Maybe you are also interested in joining. It's every Thursday, 3.30pm - 4.00pm GMT (except the first Thursday of the month, when it is shifted by half an hour to 4.00pm - 4.30pm) and takes place per zoom: https://zoom.us/j/98741582779.

meretp commented:
@jspeed-meyers Since there hasn't been any movement here for some time and we could use some feedback on our refactored version, I just wanted to ask if there is anything I can help with? 
I am also happy to add the new test files and try to fix the current issues in this branch, if that's ok with you.

We also updated the `tools-python` in the meantime so that the parser no longer fails if no creators were provided but raises an `SPDXParsingError`. This also implies that the check for `creators` 
`if len(self.doc.creation_info.creators) > 0:`
would no longer be needed here as the parser would already take care of this. This is not released yet, but just to let you know, this will need an update then.

jspeed-meyers commented:
@meretp, thank you for checking in. I would certainly appreciate some help :) 

> I am also happy to add the new test files and try to fix the current issues in this branch, if that's ok with you.

Yes, that would be more than ok, that would be lovely. Please! I'm glad to consult and discuss if you have questions or if my input would help unblock anything.

> We also updated the tools-python in the meantime so that the parser no longer fails if no creators were provided but raises an SPDXParsingError. This also implies that the check for creators
if len(self.doc.creation_info.creators) > 0:
would no longer be needed here as the parser would already take care of this. This is not released yet, but just to let you know, this will need an update then.

Thank you for the heads-up!

Also, I do want to come to the meeting you mentioned :) I was traveling last week and so I've been a little out of sorts :)

Commit:
delete unused test files

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
add valid test data from spdx-spec repo

- some elements were deleted to match the ntia minimum elements

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
update according to changes in spdx-tools

- the parser of the refactored version of spdx-tools is stricter and raises an error if the provided document doesn't contain a creator, created date or SPDXIDs for any element. We catch this error and create an individual output in this case.
- update the test data
- assert for parsing error

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
add output if the document is not valid according to the spdx specification

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
make example file for generated html output valid

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
use string instead of docstring to shorten code for html output

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

Commit:
fix linting

Signed-off-by: Meret Behrens <meret.behrens@tngtech.com>

meretp commented:
@jspeed-meyers I did some work on this and pushed the changes to the [branch on my fork](https://github.com/meretp/ntia-conformance-checker/tree/migrate2spdx-tools-0.8.0a1) as I don't have the permission to push to this branch directly. I can also open a PR against this branch, might be easier to see the changes... 

4 tests are still failing due to bugs in the `spdx-tools`. We work on fixes for these bugs and plan to do another alpha release including the fixes, so this PR should probably wait until the release and then include the second alpha release as dependency.

I wasn't sure about the generated output, especially the formatting. As the parser in the refactored version of the `spdx-tools` is stricter I decided to except the possibly thrown `SPDXParsingError` and generate an indiviudal output as all other checks can't be performed. I also added the output of validation messages if the document is parseable but not compliant with the spdx specification. Please tell me what you think about these changes.

jspeed-meyers commented:
@meretp, first, thank you! This is tremendous.

Please do open a PR against this branch. After that, I'll do a review and point out any thoughts I have. We can then wait for the next alpha release of `spdx-tools`.

Again, thank you. I hope this migration work, from myself and now you, has benefitted `spdx-tools`!

Commit:
Merge pull request #122 from meretp/migrate2spdx-tools-0.8.0a1

Migrate2spdx tools 0.8.0a1

jspeed-meyers commented:
@meretp, amazing. Thank you! What great work. I'll review this new branch now and ask any questions. I have no doubt this code is excellent.

jspeed-meyers commented:
No questions! This looks really great. I'm also of the opinion that the best way way to find bugs is through users, so even though this is a lot of change, I'm not convinced it's worth it going over it too closely in code review :) Again, great work.

You mentioned that a new future pre-release will fix the remaining bugs. I'll watch out for that. AND THANK YOU!

meretp commented:
With pleasure, thank you for the feedback and the great cooperation! This helps us a lot in the development of `spdx-tools`!

The new release is now on [Pypi](https://pypi.org/project/spdx-tools/0.8.0a2/). This should fix the tests. It seems that I've missed to format one file to also fix the black-formatting action. So if you could change the dependency to `0.8.0a2` and fix the formatting the CI should be green and this PR can be merged.

Commit:
Update spdx-tools to 0.8.0a2

Commit:
Update pipfile with 0.8.0a2 spdx-tools bump

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Merge pull request #123 from jspeed-meyers/migrate2spdx-tools-0.8.0a1

Update pipfile with 0.8.0a2 spdx-tools bump

Commit:
Perform black formatting sbom_checker.py

Signed-off-by: John Speed Meyers <jsmeyers@chainguard.dev>

Commit:
Merge pull request #124 from jspeed-meyers/migrate2spdx-tools-0.8.0a1

Perform black formatting sbom_checker.py

goneall reviewed (approved):
LGTM

jspeed-meyers commented:
Thank you, @goneall!

meretp commented on pyproject.toml:
```diff
@@ -30,7 +30,7 @@ classifiers = [
 requires-python = ">=3.8"
 keywords = ["spdx", "sbom", "ntia"]
-dependencies = [ "spdx-tools>=0.7.1"]
+dependencies = [ "spdx-tools>=0.8.0a1"]
```
```suggestion
dependencies = [ "spdx-tools>=0.8.0a2"]
```

meretp reviewed (approved):
Sorry for the late response, looks good to me! Only one small remark from my side to make the dependency consistent in the `pyoproject.toml`.

Commit:
Bump pyproject.toml spdx-tools version

Co-authored-by: Meret B. <50019307+meretp@users.noreply.github.com>

jspeed-meyers commented:
Thank you, @meretp!

jspeed-meyers commented:
Woohoo, merging! 🥳
